Scheduled Updates Add notification settings query, mutation and UI

TimBroddin opened the pull request:
Closes https://github.com/Automattic/wp-calypso/issues/89566

🚧⚠️ Do not merge until we have everything in place ⚠️🚧

## Proposed Changes

* Adds `useScheduledUpdatesNotificationSettingsQuery`
* Adds `useScheduledUpdatesNotificationSettingsMutation`
* Adds /notifications route
* Adds UI to manage notification settings


![CleanShot 2024-04-17 at 14 35 26@2x](https://github.com/Automattic/wp-calypso/assets/528287/bf340fbb-293c-4c6e-a490-298b5f88486f)
![CleanShot 2024-04-17 at 14 35 43@2x](https://github.com/Automattic/wp-calypso/assets/528287/dcc81fca-5caf-4e68-ba9c-6620e7635877)

## Testing Instructions

- Apply this PR
- Visit Scheduled Updates
- Play around

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

TimBroddin committed:
Add notification settings query, mutation and UI

TimBroddin commented:
@JanaMW27 I created this very basic UI. What do you think? Should it be aligned with the table?

bogiii commented on client/data/plugins/use-scheduled-updates-notification-settings-mutation.ts:
```diff
@@ -0,0 +1,59 @@
+	const queryClient = useQueryClient();
+
+	const mutation = useMutation( {
+		mutationKey: [ 'create-update-schedule', siteSlug ],
```
Should we rename the mutation key here to be more specific for the notification context?

TimBroddin replied:
Oops, I clearly overlooked this one.

JanaMW27 commented:
Based on our conversation had over Slack, and considering that in the future the 2 schedules limit may be lifted, I proposed to not have the entry point at the bottom of the list.

I think an approached that is more inline with Calypso interactions would be as follows;
![Email notifications](https://github.com/Automattic/wp-calypso/assets/71137829/be1923b0-a197-4c51-af4d-afba99341b7a)

[Figma file](https://www.figma.com/file/kjZaFpCQCfJnwpDdyeojNK/Plugins-scheduler?type=design&node-id=4712%3A8756&mode=design&t=6XkoDMM6ylhRPryN-1)

TimBroddin committed:
Typo

TimBroddin committed:
Add form on different route

TimBroddin committed:
Don't disable button on touched, but on pending

bogiii commented on client/blocks/plugins-scheduled-updates/notification-settings.scss:
```diff
@@ -0,0 +1,24 @@
+			font-size: $font-body-small;
+			font-weight: 500;
+			color: $studio-black;
+			margin-bottom: 8px;
```
Throughout the module, we rely on the `rem` unit for the margin and padding properties.
Please update it accordingly.

`margin-bottom: 0.5rem`

TimBroddin replied:
Thanks for the pointer, I'll update it 👍

bogiii commented on client/blocks/plugins-scheduled-updates/notification-settings.scss:
```diff
@@ -0,0 +1,24 @@
+		}
+
+		.form-field {
+			margin-bottom: 16px;
```
`margin-bottom: 1rem`

bogiii commented on client/blocks/plugins-scheduled-updates/notification-settings.scss:
```diff
@@ -0,0 +1,24 @@
+			font-size: $font-body-small;
+			color: $studio-gray-60;
+			line-height: 20px;
+			margin-bottom: 16px;
```
`margin-bottom: 1rem`

TimBroddin committed:
px -> rem

TimBroddin commented:
I'm going to disable the checkboxes and link to `me/notifications/updates` when the user has opted out of the mailing list category.

TimBroddin committed:
Disable checkboxes when opted-out globally + show help label.

TimBroddin commented:
![CleanShot 2024-04-17 at 17 12 42@2x](https://github.com/Automattic/wp-calypso/assets/528287/9cdf55a9-cc40-40c8-ada2-7ff73e018c12)

@obenland I've added a notification message for when a user has opted out of scheduled update notifications globally.

> "You have opted out of scheduled updates notifications on WordPress.com. Visit Notification Settings to enable scheduled updates notifications."

While this conveys the basic message, the phrasing could be improved to enhance clarity. Should we have someone review this copy?

obenland commented:
> I've added a notification message for when a user has opted out of scheduled update notifications globally.

I wonder if that's necessary? When they opt out, can we update the option to set everything to false?
And when they come back and enable emails again, reset the opt-out automatically?

TimBroddin commented:
> > I've added a notification message for when a user has opted out of scheduled update notifications globally.
> 
> I wonder if that's necessary? When they opt out, can we update the option to set everything to false? And when they come back and enable emails again, reset the opt-out automatically?

That would simplify it for the user. I'll took into that tomorrow!

ouikhuan commented:
> I wonder if that's necessary? When they opt out, can we update the option to set everything to false?
And when they come back and enable emails again, reset the opt-out automatically?

What happens when a user opts out of scheduled update notifications globally and then navigates to the notification settings page? How can we ensure that the global setting and the sub-settings for scheduled updates are properly synchronized and communicated to the user?

TimBroddin commented:
> > I wonder if that's necessary? When they opt out, can we update the option to set everything to false?
> > And when they come back and enable emails again, reset the opt-out automatically?
> 
> What happens when a user opts out of scheduled update notifications globally and then navigates to the notification settings page? How can we ensure that the global setting and the sub-settings for scheduled updates are properly synchronized and communicated to the user?

Good question. How about:

1. Global = off. Render checkboxes as unchecked
2. Once you select a checkbox we turn global on as well.

The only downside is: once you turn on the settings on one site, other sites will be turned on as well. 

Maybe we should just go with disabling and showing the label?

ouikhuan commented:
> The only downside is: once you turn on the settings on one site, other sites will be turned on as well.

Yeah, each site may have its own unique requirements and preferences when it comes to scheduled update notifications. Enabling the settings globally across all sites could lead to unintended consequences and potentially overwhelming users with notifications they may not want.

> Maybe we should just go with disabling and showing the label?

I'd go with this approach too, it effectively communicates the global opt-out state to the user while preserving the individual site settings, but happy to discuss if there's a better way to address this problem.

ouikhuan commented on client/blocks/plugins-scheduled-updates/index.tsx:
```diff
@@ -98,16 +105,30 @@ export const PluginsScheduledUpdates = ( props: Props ) => {
-						>
-							{ translate( 'Add new schedule' ) }
-						</Button>
+					{ context === 'list' && ! hideButtons && (
```
If there are two schedules, the Notification management button won't show up in this case.

TimBroddin replied:
Nice catch. I'll fix it.

ouikhuan commented on client/blocks/plugins-scheduled-updates/index.tsx:
```diff
@@ -98,16 +105,30 @@ export const PluginsScheduledUpdates = ( props: Props ) => {
+									variant="secondary"
+									onClick={ onNotificationManagement }
+								>
+									{ translate( 'Notification management' ) }
```
Should we call it Notification Settings instead?

ouikhuan commented on client/blocks/plugins-scheduled-updates/notification-settings.tsx:
```diff
@@ -0,0 +1,125 @@
+						</Button>
+					) }
+				</div>
+				<Text>{ translate( 'Notification management' ) }</Text>
```
Should we call it Notification Settings instead?

ouikhuan commented on client/blocks/plugins-scheduled-updates/index.tsx:
```diff
@@ -84,6 +87,10 @@ export const PluginsScheduledUpdates = ( props: Props ) => {
 		},
+		notifications: {
+			component: <NotificationSettings onNavBack={ onNavBack } />,
+			title: translate( 'Notification management' ),
```
Should we call it Notification Settings instead?

ouikhuan commented on client/blocks/plugins-scheduled-updates/notification-settings.tsx:
```diff
@@ -0,0 +1,125 @@
+					/>
+				</div>
+
+				{ ! wpcomNotificationSettings?.scheduled_updates && (
```
This message shows on loading and disappear later, can we wait until this is fetched before showing the message?

TimBroddin committed:
- Copy changes
- Enable checkboxes during loading of global settings
- Show button when there are 2 schedules

ouikhuan commented on client/blocks/plugins-scheduled-updates/notification-settings.tsx:
```diff
@@ -0,0 +1,126 @@
+						label={ translate( 'On successful updates' ) }
+						checked={ formValues.success }
+						onChange={ handleCheckboxChange( 'success' ) }
+						disabled={ ! isFetched || ! hasGlobalNotificationsDisabled }
```
Should this be `hasGlobalNotificationsDisabled` instead?

TimBroddin committed:
Typo

TimBroddin committed:
Show success/failure message on save

ouikhuan reviewed (approved):
LGTM, thanks for addressing all the feedback 😄

TimBroddin commented:
@michaelpick Could you (or someone else from Content) take a quick look at this sentence?

"You have opted out of scheduled updates notifications on WordPress.com. Visit {{notificationSettingsLink}}Notification Settings{{/notificationSettingsLink}} to enable scheduled updates notifications."

The context is that we disable the form to edit your scheduled updates notification settings, when you've unsubscribed from such notifications globally.

![image](https://github.com/Automattic/wp-calypso/assets/528287/613af1f4-e3ae-4cd7-bd62-408d93c9d845)

michaelpick commented:
Hey @TimBroddin 👋 

We whipped up a few alternative options for you!

- You've opted out of notifications for scheduled WordPress.com updates. [Enable].
- You've opted out of scheduled update notifications, but you can [re-enable them] at any time.
- You've opted out of scheduled update notifications on WordPress.com. Want to [re-enable them]?
- You've opted out of WordPress.com's scheduled update notifications. Need to [re-enable them]? 
- 
You've opted out of WordPress.com's scheduled update notifications. Head to [Notification Settings] to re-enable them.


A few more variations of the second line:

- You can switch them back on at any time by heading [here].
- 
You can bring them back any time from [Notifications Settings].

See what you think and let me know if you need anything else!

TimBroddin commented:
> Hey @TimBroddin 👋
> 
> We whipped up a few alternative options for you!
> 
> * You've opted out of notifications for scheduled WordPress.com updates. [Enable].
> * You've opted out of scheduled update notifications, but you can [re-enable them] at any time.
> * You've opted out of scheduled update notifications on WordPress.com. Want to [re-enable them]?
> * You've opted out of WordPress.com's scheduled update notifications. Need to [re-enable them]?
> 
> You've opted out of WordPress.com's scheduled update notifications. Head to [Notification Settings] to re-enable them.
> 
> A few more variations of the second line:
> 
> * You can switch them back on at any time by heading [here].
> 
> You can bring them back any time from [Notifications Settings].
> 
> See what you think and let me know if you need anything else!

Awesome! Thanks a lot!

TimBroddin committed:
Update copy
